Disable auto-gen when a rule has mixed of kinds: pod & pod controllers

realshuting opened the pull request:
Signed-off-by: Shuting Zhao <shutting06@gmail.com>

## Related issue

Closes https://github.com/kyverno/kyverno/issues/1805.

<!--
Please link the GitHub issue this pull request resolves in the format of `#1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## What type of PR is this
> /kind bug

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->

## Proposed Changes
When there are mixed kinds (Pod & Pod controllers) defined in `match.resources.kinds` or `exclude.resources.kinds`,  the auto-gen is disabled.

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests
Create the following policy:
```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: block-old-flux
spec:
  validationFailureAction: enforce
  background: false
  rules:
  - name: block-old-flux
    match:
      resources:
        kinds:
        - Deployment
        - CronJob
        - Job
        - StatefulSet
        - DaemonSet
        - Pod
    validate:
      message: Cannot use old Flux v1 annotation.
      pattern:
        metadata:
          =(annotations):
            X(fluxcd.io/*): "*?"
```
The auto-gen is automatically disabled by `pod-policies.kyverno.io/autogen-controllers: none`:
```yaml
✗ kg cpol block-old-flux -o yaml | k neat
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  annotations:
    pod-policies.kyverno.io/autogen-controllers: none
  name: block-old-flux
spec:
  background: false
  rules:
  - match:
      resources:
        kinds:
        - Deployment
        - CronJob
        - Job
        - StatefulSet
        - DaemonSet
        - Pod
    name: block-old-flux
    validate:
      message: Cannot use old Flux v1 annotation.
      pattern:
        metadata:
          =(annotations):
            X(fluxcd.io/*): '*?'
  validationFailureAction: enforce
```



<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [] My PR contains new or altered behavior to Kyverno and
  - [] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the doc update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->
  - [] I have read the [PR documentation guide](pr_documentation.md) and followed the process including adding proof manifests to this PR.

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

JimBugwadia commented on pkg/policymutation/policymutation.go:
```diff
@@ -224,8 +224,8 @@ func GeneratePodControllerRule(policy kyverno.ClusterPolicy, log logr.Logger) (p
 
```
missing "// scenario C"?

realshuting replied:
C is handled in https://github.com/kyverno/kyverno/pull/1847/files/2704e40165e020a3b846cf477108220d8ce24e22#diff-15f5667117ad6d93290286aaa292507f75cc3aa9441f9b8f03bdfe1cd31cd6acR320.

JimBugwadia replied:
I meant in the comments....

realshuting replied:
The comment was added to another place, not so obvious to see:
https://github.com/kyverno/kyverno/blob/2704e40165e020a3b846cf477108220d8ce24e22/pkg/policymutation/policymutation.go#L320-L321

JimBugwadia commented on pkg/policymutation/policymutation.go:
```diff
@@ -235,7 +235,7 @@ func GeneratePodControllerRule(policy kyverno.ClusterPolicy, log logr.Logger) (p
 
 	// scenario B
 	if controllers == "none" {
```
Would it be cleaner to introduce a "canAutoGen" method that returns a true / false based on various checks? 

Seems a bit confusing that getControllers returns "none" when there are controllers in the match  / exclude....

realshuting replied:
Updated the name, also returned a bool to indicate the result. We still need "controllers" as we use that to generate annotation JSON patches.

JimBugwadia reviewed (commented):
One additional question - what happens if the user sets the annotation but there are multiple controllers? Do we need a validation check for that? Any other auto-gen related validation checks we need to add?

realshuting commented:
>what happens if the user sets the annotation but there are multiple controllers?

Good catch! I added the logic to handle this scenario automatically, when the predefined controllers are invalid, Kyverno overwrites it to "none". Added tests:
https://github.com/kyverno/kyverno/blob/ee6533488ebb523fa96572db07d4c69aab45db7f/pkg/policymutation/policymutation_test.go#L222-L229

realshuting commented on pkg/policy/validate_test.go:
```diff
@@ -1293,39 +1293,24 @@ func Test_checkAutoGenRules(t *testing.T) {
 		expectedResult bool
```
The tests added in the [previous](https://github.com/kyverno/kyverno/pull/1839) PR are duplicated, updated to cover other scenarios.
